### PR TITLE
examples: add `chroot-batch`

### LINKED ISSUE #907: `build`: improve unattended builds
The current situation of having unattended builds of many packages with a build user (as documented in `aur-build(1)` and `aur-chroot(1)`) is unsatisfactory. The setup suffers from file permission conflicts between the main/current user and the build user, different sync and seen caches and potentially different `AUR_PAGER` setups, which would all need to be synced or worked around/ignored.

An alternative to the separate build user would be to increase or disable the sudo authentication timeout, potentially temporarily for the build duration, which is probably better to work with (and what I will consider picking up for now) but not ideal either because it's more general in scope than a dedicated build user with appropriate configuration.

---

The simplest approach to this is probably to run `aur` commands itself with elevated permissions, but that's sort of the "golden hammer" solution and some de-elevation of privileges is needed on 3 to 4 occasions (depending on package signing):

- https://github.com/AladW/aurutils/blob/master/lib/aur-build#L308-L313 for the packagelist
- https://github.com/AladW/aurutils/blob/master/lib/aur-build#L353-L355 for pkgver
- https://github.com/AladW/aurutils/blob/master/lib/aur-build#L387-L388 for the actual makepkg command
- gpg for signing

quoting @AladW:
> In short, to make a root `aur-build` work 
> 1. drop permissions for user commands (with `setpriv`) instead of elevating permissions for root commands 
> 2. replace `makepkg -s` (which elevates permissions) with `makepkg --printsrcinfo` (as user, to retrieve a list of dependencies) and `pacinstall` (as root) 
> 3. avoid prefixing commands with `setpriv` or `sudo` by redefining commands (as is already done for `$PACMAN_AUTH`, `$MAKEPKG`), if a build user is specified on the command-line (for example with `-U <build_user>`)

Another option (suggested by @Earnestly) would be a daemon that performs tasks that require permissions via IPC.

Related: #852

---

*Discussed on IRC from 2022-02-05 14:51 UTC+1 to 18:15 UTC+1.*

### PR DESCRIPTION
**Pull request by AladW**
Issue #907

### CODE REVIEW

**Comment by AladW** [comment]
Following https://github.com/AladW/aurutils/issues/907#issuecomment-1070958818 I would also add `aur-sync --build-user`,  but something like:
```bash
# Specify user for aur-build (#907)
if [[ -v build_user ]]; then
    ninja() {
        sudo -u "$build_user" ninja "$@"
    }
    aur() { # only applies to build--ninja, build below
        sudo -u "$build_user" aur "$@"
    }
fi

# Build dependency tree with ninja (#908)
if (( AUR_SYNC_USE_NINJA )); then
    ...
```
results in 3 `sudo` calls with `AUR_SYNC_USE_NINJA=1`, namely
```
sudo -u "$build_user" aur sync--ninja "$AURDEST" >build.ninja
sudo -u "$build_user" env NINJA_STATUS='[%s/%t] ' ninja -k "$keep_going"
sudo -u "$build_user" env NINJA_STATUS='[%s/%t] ' ninja -nC /var/empty -f "$PWD"/build.ninja
```
i.e. potentially 3 prompts instead of 1.

File: examples/chroot-batch [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+repo_add_and_sign() {
+	# ensure the passphrase is not stored as an environment variable
+	# by saving it in a local variable inside a function
+	local passphrase pkglist pkgbase
+	read -p 'GPG passphrase for signing packages: ' passphrase
+
+	#global var_tmp build_user db_path
+	for pkgbase in "$@"; do
+		cd "$startdir"
```
Comment (Earnestly) [inline examples/chroot-batch:57]: Perhaps `cd -- "$startdir"/"$pkgbase"`?

File: examples/chroot-batch [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		cd "$pkgbase"
+		PKGDEST="$var_tmp" aur chroot --build --cargs -c,-u # makechrootpkg -cu
+
+		cd "$var_tmp"
+		pkglist=(*) siglist=()
```
Comment (Earnestly) [inline examples/chroot-batch:62]: May need `nullglob` here.

File: examples/chroot-batch (hunk 1 above)
Comment (AladW) [reply examples/chroot-batch:57]: cf. c7964918fd9f82931c2c6ee2af6b7153206acbda

File: examples/chroot-batch (hunk 2 above)
Comment (AladW) [reply examples/chroot-batch:62]: I'm trying to think of a scenario where `pkglist` would be empty, but generally yes.

**Comment by AladW** [comment]
The assumption `dirname == pkgbase` could be dropped if the `--dry-run` output is null-separated (there are two relevant paths, one for the package in the local repository, the other for the `PKGBUILD` directory). This assumption is fulfilled when using `aur-sync` and `aur-fetch`, so I'll leave this aside for now.

**Pull request merged**